File: 234-PalindromeLinkedList/234-PalindromeLinkedList.py
# Last updated: 8/16/2026, 9:51:00 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # converted_list = []
        # while head:
        #     converted_list.append(head.val)
        #     head = head.next
        
        # l = 0
        # r = len(converted_list) - 1
        # while l <= r:
        #     if converted_list[l] != converted_list[r]:
        #         return False
        #     l += 1
        #     r -= 1
        
        # return True  

        # method-2: reverse the second half of the linked list, and then compare
        # Floyd's Cycle to find the middle point
        # when fast reades end (fast == null || fast.next == null), slow is at MIDDLE

        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        # slow is the middle point, reverse it
        reverse = self.reverse(slow)

        while reverse:
            if reverse.val != head.val: return False
            reverse = reverse.next
            head = head.next
        
        return True


    # reverse linked list starting with 'head'
    def reverse(self, head: ListNode) -> ListNode:
        prev = None
        curr = head
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev

File: 234-PalindromeLinkedList/234-PalindromeLinkedList.java
// Last updated: 8/16/2026, 9:51:03 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        // // linkedList is queue, FIFO
        // // Stack, LIFO
        // // comapre queue and stack one by one -> palindrome order

        // Stack<Integer> stack = new Stack<>();
        // ListNode curr = head;
        // while( curr != null) {
        //     stack.push(curr.val);
        //     curr = curr.next;
        // }
        
        // while(head != null) {
        //     if(head.val != stack.peek()) {
        //         return false;
        //     } else {
        //         head = head.next;
        //         stack.pop();
        //     }
        // }
        // return true;

        // method 2 - Floyd's Cycle, check marked solution for expalnation
        // when fast reades end (fast == null || fast.next == null), slow is at MIDDLE
        // reverse the back half of the list with a 'prev' node,
        // so they points to prev, instead of next
        // after reversing, slow is at the end
        // restart fast from the begin
        // compare begin vs. end
        if(head == null || head.next == null) return true;
        ListNode fast = head, slow = head;
        ListNode prev, temp;
        while(fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        
        // reverse the back half of linkedList
        prev = slow;
        slow = slow.next;
        prev.next = null; // middle points to null, avoid cycle (endless loop)
        while(slow != null) {
            temp = slow.next;
            slow.next = prev;
            // move to next nodes
            prev = slow;
            slow = temp;
        }

        // slow is at the end (null), reset fast, compare slow and fast
        fast = head;
        slow = prev; // (prev points to the last node, slow is null)
        while(slow != null) {
            if(fast.val != slow.val) return false;
            fast = fast.next;
            slow = slow.next;
        }
        return true;
    }
}